compiler/noirc_frontend/src/hir/def_collector/dc_crate.rs

```diff
@@ -386,22 +424,117 @@ fn collect_impls(
 fn collect_trait_impls(
     context: &mut Context,
     crate_id: CrateId,
-    collected_impls: &TraitImplMap,
+    collected_impls: &mut [UnresolvedTraitImpl],
 ) -> Vec<(CompilationError, FileId)> {
     let interner = &mut context.def_interner;
     let def_maps = &mut context.def_maps;
     let mut errors: Vec<(CompilationError, FileId)> = vec![];
 
-    // TODO: To follow the semantics of Rust, we must allow the impl if either
-    //     1. The type is a struct and it's defined in the current crate
-    //     2. The trait is defined in the current crate
-    for ((unresolved_type, module_id, _), trait_impl) in collected_impls {
-        let path_resolver =
-            StandardPathResolver::new(ModuleId { local_id: *module_id, krate: crate_id });
+    for trait_impl in collected_impls.iter_mut() {
+        let module_id = trait_impl.module_id;
+        let unresolved_type = trait_impl.object_type.clone();
+
+        let module = ModuleId { local_id: module_id, krate: crate_id };
+        let resolve_result = resolve_trait_by_path(def_maps, module, trait_impl.trait_path.clone());
+
+        trait_impl.the_trait = match resolve_result {
+            Ok(trait_id) => Some(trait_id),
+            Err(error) => {
+                errors.push((error.into(), trait_impl.file_id));
+                None
+            }
+        };
+
+        if let Some(trait_id) = trait_impl.the_trait {
+            let the_trait = interner.get_trait(trait_id);
+            let the_trait = the_trait.borrow();
```

Comment:
Fix: https://github.com/noir-lang/noir/commit/78cff10dce4ded304a49de818e7bf00d5de2f66f

I'm hesitant about this change, I know you're serious about avoiding unnecessary clones and removing the `Shared` adds a very expensive one in `get_trait` - Trait is currently 300B + 150B for every TraitFunction + a bunch of Vecs and Idents that will all allocate, and we're cloning all that just to usually iterate the functions.

Returning the trait by reference in `get_trait` is not really an options since then we can't do this:
```rust
let the_trait: &Trait = interner.get_trait(id);
for fun in the_trait.functions {
    // can't use mutable interner here since its partially borrowed :/
}
```

If you know of some way to avoid this without Shared I'm open to suggestions

Reply:
That's a good point. I was originally going to suggest returning it by reference and if the interner needs to be mutated to drop the reference and re-acquire it again afterward with another call to `get_trait`. You may be able to do this still and collect any changes in the for loop in a separate Vec if needed and apply them afterward, but at that point it's not much of an improvement over the Shared wrapper. I probably made the wrong call here then, so we can revert it back to the Shared in a later PR.